pkg/informer/informer.go

```diff
@@ -212,42 +260,78 @@ func (d *DynamicDiscoverySharedInformerFactory) discoverTypes(ctx context.Contex
 					continue
 				}
 
-				latest.Insert(strings.Join([]string{ai.Name, gv.Version, gv.Group}, "."))
+				latest[gvr] = struct{}{}
 			}
 		}
 	}
 
+	// Grab a read lock to compare against d.gvrs to see if we need to start or stop any informers
+	d.mu.RLock()
+	var informersToAdd []schema.GroupVersionResource
+	for gvr := range latest {
+		if _, found := d.gvrs[gvr]; !found {
+			informersToAdd = append(informersToAdd, gvr)
+		}
+	}
+
+	var informersToRemove []schema.GroupVersionResource
+	for gvr := range d.gvrs {
+		if _, found := latest[gvr]; !found {
+			informersToRemove = append(informersToRemove, gvr)
+		}
+	}
+	d.mu.RUnlock()
+
+	if len(informersToAdd) == 0 && len(informersToRemove) == 0 {
+		return nil
+	}
+
+	// We have to add/remove, so we need the write lock
```

Comment:
Isn't there a race here in case someone else grabbed the write lock first and mucked with things since we last read?

Reply:
Will ponder

Reply:
Fixed